src/utils/map-state.js

```diff
@@ -0,0 +1,332 @@
+import {PerspectiveMercatorViewport} from 'viewport-mercator-project';
+import assert from 'assert';
+
+// MAPBOX LIMITS
+const MAX_PITCH = 60;
+const MAX_ZOOM = 20;
+
+function mod(value, divisor) {
+  const modulus = value % divisor;
+  return modulus < 0 ? divisor + modulus : modulus;
+}
+
+const propTypes = {
+  /** The width of the map */
+  width: PropTypes.number.isRequired,
+  /** The height of the map */
+  height: PropTypes.number.isRequired,
+  /** The latitude of the center of the map. */
+  latitude: PropTypes.number.isRequired,
+  /** The longitude of the center of the map. */
+  longitude: PropTypes.number.isRequired,
+  /** The tile zoom level of the map. */
+  zoom: PropTypes.number.isRequired,
+  /** Specify the bearing of the viewport */
+  bearing: PropTypes.number,
+  /** Specify the pitch of the viewport */
+  pitch: PropTypes.number,
+  /**
+    * Specify the altitude of the viewport camera
+    * Unit: map heights, default 1.5
+    * Non-public API, see https://github.com/mapbox/mapbox-gl-js/issues/1137
+    */
+  altitude: PropTypes.number,
+
+  /** Constraints */
+  maxZoom: PropTypes.number,
+  minZoom: PropTypes.number,
+  maxPitch: PropTypes.number,
+  minPitch: PropTypes.number,
+
+  /**
+    * Required to calculate the mouse projection during panning.
+    * The point on map being grabbed when the operation first started.
+    */
+  startPanLngLat: PropTypes.arrayOf(PropTypes.number),
+  /**
+    * Required to calculate the mouse projection during zooming.
+    * Center of the zoom when the operation first started.
+    */
+  startZoomLngLat: PropTypes.arrayOf(PropTypes.number),
+  /** Bearing when current perspective rotate operation started */
+  startBearing: PropTypes.number,
+  /** Pitch when current perspective rotate operation started */
+  startPitch: PropTypes.number,
+  /** Zoom when current zoom operation started */
+  startZoom: PropTypes.number
+};
+
+export default class MapState {
+
+  constructor({
+    /** Mapbox viewport properties */
+    width,
+    height,
+    latitude,
+    longitude,
+    zoom,
+    bearing = 0,
+    pitch = 0,
+    altitude = 1.5,
+
+    /** Viewport constraints */
+    maxZoom = MAX_ZOOM,
+    minZoom = 0,
+    maxPitch = MAX_PITCH,
+    minPitch = 0,
+
+    /** Interaction states */
+    startPanLngLat,
+    startZoomLngLat,
+    startBearing,
+    startPitch,
+    startZoom
+  } = {}) {
+    this.props = {
+      width,
+      height,
+      latitude,
+      longitude,
+      zoom,
+      bearing,
+      pitch,
+      altitude,
+      maxZoom,
+      minZoom,
+      maxPitch,
+      minPitch,
+      startPanLngLat,
+      startBearing,
+      startPitch
+    };
+  }
+
+  _updateViewport(opts) {
+    // Update props
+    Object.assign(this.props, opts);
+    this._applyConstraints();
+    return this;
+  }
+
+  // Apply any constraints (mathematical or defined by props) to viewport params
+  _applyConstraints() {
+    const viewport = this.props;
```

Comment:
We are mutating props here, is that desirable? 

Should we comment in that case, explain why this is OK.

Should we take the props as an argument instead of using `this.props`?